### ISSUE TITLE
Implement a user model with three access tiers

### ISSUE THREAD

**Issue by tskir** [EBIvariation/trait-curation#75]
There should be three access levels:

- Admin (using Django standard field): can do anything. Only admin can view the configuration (admin) page #50.
- Staff (also using Django standard field): can comment, propose and review mappings.
- User who is not logged in, or logged in but does not have any of the above roles: can only view the pages (not the admin page).

**Comment by joj0s** [comment]
I'd like to point out here that I was a bit mistaken when describing the default user fields. A user with the `is_staff` attribute *can* use the admin site, this is what this field suggests. And a user with the `is_superuser` attribute, as the name suggests, has complete access in the app, meaning all permissions, and can change other user's status.

If we want to go with the 3 tier approach, then I suggest we remove the 'role' field we added, since we can implement 3 tier access using the current model's default auth fields entirely.

**Comment by tskir** [comment]
> If we want to go with the 3 tier approach, then I suggest we remove the 'role' field we added, since we can implement 3 tier access using the current model's default auth fields entirely.

Yes, this was my idea as well. It is enticing to get rid of a field and just use built-in functionality.

The only thing which concerns me is that **is_staff** users (which will be the regular curators and reviewers) are able to use the admin site. Does this mean they can in theory edit any database entries directly? Because they shouldn't be able to, and this should be reserved only for superusers (admins).

**Comment by joj0s** [comment]
> The only thing which concerns me is that is_staff users (which will be the regular curators and reviewers) are able to use the admin site. Does this mean they can in theory edit any database entries directly? 

Before we decide that, I would like to point out that the admin site is completely optional, and although it is really useful during development, we can completely forego it during production. So first of all we should consider whether we will keep it or not.

**Comment by tskir** [comment]
It also looks like access to the admin site and editing the entities can also be controlled for staff users with high degree of granularity: https://stackoverflow.com/a/20470185

So, let's go with this approach: abandon the "role" attribute, and use the built-in functionality and roles.

**Comment by joj0s** [comment]
Yes so let's sum up what I found about the built-in auth functionality:

**There are two flags that are used to control permissions for users, non mutually exclusive:**

- *is_superuser:* It means that by default, the user has all permissions in the app. These permissions means he can perform the usual CRUD operations on any resource of the application.

- *is_staff:* This solely controls whether the user can log in to the admin site. 

Important notes: 
- We can specify for each model, what permissions we want the staff to have on it. So if we want for example the staff to be able to update ontology terms via the admin site, we can do that by giving them the permission in the data model.

- A non-superuser staff, will only see an empty admin site, because he cannot explicitly perform CRUD operations in any model by default.

- A non-staff but superuser staff, cannot log in to the admin site. Of course we would want to give superusers both permissions.

**Comment by tskir** [comment]
Thanks, that's very useful! Sounds like we can make curators and reviewers to be non-superuser staff, and to make admins superuser staff. But the main question here is: can we use the "staff" flag to change the display of pages and authorising the requests for the app itself (not for the admin site)? Because we **want** our non-superuser staff to be able to do curations and reviews using the app pages, but we **do not want** them be able to modify anything at all in the model/admin site directly.

**Comment by joj0s** [comment]
> can we use the "staff" flag to change the display of pages and authorising the requests for the app itself

I am quite positive we can. Again, only superusers have permission to see the models in the admin site, and we can use the `is_staff` flag for own purposes to do manual checks for it.

**Comment by tskir** [comment]
OK great, then I'm happy with everything and we can go this way

**Comment by joj0s** [comment]
> User who is not logged in, or logged in but does not have any of the above roles: can only view the pages (not the admin page).

Just to clarify, these roles are considered equal right?

Now regarding the staff role, should I somehow assign that automaically? I am thinking that if a user successfully logs in with an EBI email, I can assign him the staff role.

**Comment by tskir** [comment]
> Just to clarify, these roles are considered equal right?

Yes, "not logged in" and "not logged in without any roles" is the same.

> Now regarding the staff role, should I somehow assign that automaically? I am thinking that if a user successfully logs in with an EBI email, I can assign him the staff role.

I don't think this is necessary at this point, we can just have some hardcoded admin email (e.g. mine?) and then this person can assign other admins and staff members
